File: internal/cli/audit.go
package cli

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"strings"
	"time"

	"github.com/spf13/cobra"

	"github.com/starfield17/rhost/internal/audit"
	"github.com/starfield17/rhost/internal/config"
	"github.com/starfield17/rhost/internal/errs"
	"github.com/starfield17/rhost/internal/output"
)

// auditTimer records one operation when it finishes (docs/ARCHITECTURE.md §36).
//
// Auditing is fail-open: a write error is reported on stderr and the operation
// is unaffected, because a full or read-only disk must not make remote work
// impossible. When auditing is disabled (RHOST_AUDIT=0) the timer is inert, so a
// call site never branches on whether it is on.
type auditTimer struct {
	start time.Time
	rec   *audit.Recorder
	op    string
	host  string
}

func startAudit(op, host string) *auditTimer {
	t := &auditTimer{start: time.Now(), op: op, host: host}
	if audit.Enabled() {
		t.rec = &audit.Recorder{Path: audit.Path(config.StateDir())}
	}
	return t
}

// succeed records a completed operation. cwd and command may be empty; exitCode
// is nil for operations that have no remote status of their own.
func (t *auditTimer) succeed(cwd, command string, exitCode *int) {
	t.write(audit.Entry{OK: true, Cwd: cwd, Command: audit.Summarize(command), ExitCode: exitCode})
}

// fail records a refused or failed operation.
func (t *auditTimer) fail(aerr *errs.Error) {
	e := audit.Entry{OK: false}
	if aerr != nil {
		e.ErrorCode = string(aerr.Code)
	}
	t.write(e)
}

func (t *auditTimer) write(e audit.Entry) {
	if t == nil || t.rec == nil {
		return
	}
	e.Time = time.Now().UTC().Format(time.RFC3339)
	e.Operation = t.op
	e.Host = t.host
	e.DurationMS = time.Since(t.start).Milliseconds()
	if err := t.rec.Record(e); err != nil {
		fmt.Fprintf(os.Stderr, "rhost: audit: %v\n", err)
	}
}

// auditListing is the JSON payload of `rhost audit`.
type auditListing struct {
	Path    string        `json:"path"`
	Entries []audit.Entry `json:"entries"`
}

func newAuditCmd() *cobra.Command {
	var (
		limit int
		host  string
	)
	cmd := &cobra.Command{
		Use:   "audit",
		Short: "Read the local audit log of remote operations",
		Long: `Read rhost's local audit log — a JSON Lines file of the operations this
machine has run against remote hosts (docs/ARCHITECTURE.md §36).

It records operations, not secrets: no environment maps and no file contents.
Logging is fail-open and can be turned off with RHOST_AUDIT=0.

By default the most recent 20 entries are shown; --limit 0 shows all.`,
		Args: cobra.NoArgs,
		RunE: func(cmd *cobra.Command, _ []string) error {
			path := audit.Path(config.StateDir())
			entries, aerr := readAuditLog(path)
			if aerr != nil {
				emitFailure("audit", "", aerr)
				return nil
			}
			if host != "" {
				entries = filterAuditHost(entries, host)
			}
			if limit > 0 && len(entries) > limit {
				entries = entries[len(entries)-limit:]
			}
			if jsonFlag {
				_ = output.Success("audit", "", auditListing{Path: path, Entries: entries}).Write(os.Stdout)
				return nil
			}
			renderAudit(os.Stdout, path, entries)
			return nil
		},
	}
	cmd.Flags().IntVar(&limit, "limit", 20, "show at most the last N entries (0 = all)")
	cmd.Flags().StringVar(&host, "host", "", "only entries for this host")
	return cmd
}

// readAuditLog parses the JSONL file. A missing file is an empty log, not an
// error; a corrupt line is skipped rather than making the whole trail unreadable,
// which matters because the log is appended to by other processes concurrently.
func readAuditLog(path string) ([]audit.Entry, *errs.Error) {
	f, err := os.Open(path)
	if os.IsNotExist(err) {
		return []audit.Entry{}, nil
	}
	if err != nil {
		return nil, errs.Wrap(errs.ConfigInvalid, "cannot read audit log: "+err.Error(), false, err)
	}
	defer f.Close()

	out := []audit.Entry{}
	sc := bufio.NewScanner(f)
	sc.Buffer(make([]byte, 0, 64*1024), 1024*1024)
	for sc.Scan() {
		line := strings.TrimSpace(sc.Text())
		if line == "" {
			continue
		}
		var e audit.Entry
		if err := json.Unmarshal([]byte(line), &e); err != nil {
			continue
		}
		out = append(out, e)
	}
	if err := sc.Err(); err != nil {
		return nil, errs.Wrap(errs.ConfigInvalid, "reading audit log: "+err.Error(), false, err)
	}
	return out, nil
}

func filterAuditHost(entries []audit.Entry, host string) []audit.Entry {
	out := []audit.Entry{}
	for _, e := range entries {
		if e.Host == host {
			out = append(out, e)
		}
	}
	return out
}

func renderAudit(w io.Writer, path string, entries []audit.Entry) {
	if len(entries) == 0 {
		fmt.Fprintf(w, "no audit entries (%s)\n", path)
		return
	}
	for _, e := range entries {
		status := "ok"
		if !e.OK {
			status = e.ErrorCode
			if status == "" {
				status = "failed"
			}
		}
		fmt.Fprintf(w, "%s  %-10s %-18s %-8s %s\n", e.Time, dash(e.Host), e.Operation, status, e.Command)
	}
}

func dash(s string) string {
	if s == "" {
		return "-"
	}
	return s
}

File: src/cli/audit.rs
//! The `audit` command, and the timer the other commands use to write to it.
//!
//! Auditing is fail-open: a write that does not land is reported on stderr and the
//! operation it describes is unaffected, because a full or read-only disk must not
//! make remote work impossible. `RHOST_AUDIT=0` makes the timer inert, so no call
//! site branches on whether auditing is on.

use super::console::{Sink, warn};
use super::grammar::{FlagSpec, JSON, help_or_error, parse, usage_error};
use super::{Command, Help, Scope};
use crate::audit;
use crate::config;
use crate::output;
use std::time::Instant;

/// The flags `audit` takes. It reads a local file, so it has no host operand.
pub(crate) static AUDIT_FLAGS: &[FlagSpec] = &[
    JSON,
    FlagSpec::value("limit", false, "show at most N recent entries; 0 shows all"),
    FlagSpec::value("host", false, "only entries for this host"),
];

/// How many entries a caller who said nothing gets. The trail is append-only and
/// unbounded, so the default view is the tail.
const DEFAULT_LIMIT: i64 = 20;

pub(crate) fn command(argv: &[String], json: bool) -> Command {
    let parsed = match parse(argv, AUDIT_FLAGS) {
        Ok(parsed) => parsed,
        Err(message) => return usage_error(Scope::Root, message),
    };
    if parsed.bool_flag("help") {
        return help_or_error(Scope::Root, Help::Audit, json);
    }
    if !parsed.operands.is_empty() || parsed.dash {
        return usage_error(
            Scope::Root,
            format!(
                "rhost audit accepts 0 arg(s), received {}",
                parsed.operands.len()
            ),
        );
    }
    let limit = match parsed.value("limit") {
        None => DEFAULT_LIMIT,
        Some(raw) => match raw.parse::<i64>() {
            Ok(value) if value >= 0 => value,
            _ => {
                return usage_error(
                    Scope::Root,
                    format!("--limit must be zero or a positive integer, not {raw:?}"),
                );
            }
        },
    };
    Command::Audit {
        limit: limit as usize,
        host: parsed.value("host").unwrap_or("").to_string(),
    }
}

pub(crate) fn run(sink: &mut Sink, limit: usize, host: &str, json: bool) -> u8 {
    let path = audit::path(&config::v4_state_dir());
    let entries = match audit::read(&path) {
        Ok(entries) => entries,
        Err(error) => {
            return super::console::Failure::new(
                "audit",
                "",
                "CONFIG_INVALID",
                format!("cannot read audit log: {error}"),
            )
            .deliver(sink, json);
        }
    };
    let entries = if host.is_empty() {
        entries
    } else {
        audit::filter_host(entries, host)
    };
    let entries = audit::keep_last(entries, limit);
    if json {
        sink.envelope(&output::audit(&path, &entries));
        return 0;
    }
    if entries.is_empty() {
        sink.line(&format!("no audit entries ({})", path.display()));
        return 0;
    }
    for entry in &entries {
        let status = if entry.ok {
            "ok".to_string()
        } else if entry.error_code.is_empty() {
            "failed".to_string()
        } else {
            entry.error_code.clone()
        };
        sink.line(&format!(
            "{}  {:<10} {:<18} {:<8} {}",
            entry.time,
            dash(&entry.host),
            entry.operation,
            status,
            entry.command_summary
        ));
    }
    0
}

fn dash(text: &str) -> &str {
    if text.is_empty() { "-" } else { text }
}

/// One operation, timed, and written to the trail when the caller says how it
/// ended. A recorder that cannot write never fails the operation.
pub(crate) struct Timer {
    start: Instant,
    recorder: audit::Recorder,
    operation: &'static str,
    host: String,
}

impl Timer {
    pub(crate) fn start(operation: &'static str, host: &str) -> Self {
        Self {
            start: Instant::now(),
            recorder: audit::Recorder::from_env(&config::v4_state_dir()),
            operation,
            host: host.to_string(),
        }
    }

    /// Records an operation that completed as designed. `exit_code` is the remote
    /// command's own status when it has one, and `None` for operations whose
    /// answer is not a status.
    pub(crate) fn succeeded(&self, cwd: &str, command_summary: &str, exit_code: Option<u8>) {
        self.write(&audit::Entry::completed(
            self.operation,
            &self.host,
            cwd,
            command_summary,
            exit_code,
            self.elapsed_ms(),
        ));
    }

    /// Records a refusal or a failure, keeping the code an agent would branch on.
    pub(crate) fn failed(&self, error_code: &str) {
        self.write(&audit::Entry::refused(
            self.operation,
            &self.host,
            error_code,
            self.elapsed_ms(),
        ));
    }

    fn elapsed_ms(&self) -> u64 {
        u64::try_from(self.start.elapsed().as_millis()).unwrap_or(u64::MAX)
    }

    fn write(&self, entry: &audit::Entry) {
        if let Err(error) = self.recorder.record(entry) {
            warn(&format!("rhost: audit: {error}"));
        }
    }
}

/// Records one entry the caller has already timed and described, used where one
/// command covers several operations (a batch) and each one has its own duration.
pub(crate) fn record(entry: audit::Entry) {
    let recorder = audit::Recorder::from_env(&config::v4_state_dir());
    if let Err(error) = recorder.record(&entry) {
        warn(&format!("rhost: audit: {error}"));
    }
}
